File: native/ZiNotes/app/src/main/java/com/example/zinotes/ui/DetailScreen.kt
package com.example.zinotes.ui

import androidx.compose.animation.Crossfade
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.IconButtonDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewmodel.compose.viewModel
import com.example.zinotes.R
import com.example.zinotes.state.DetailUiState
import com.example.zinotes.room.Hanzi
import com.example.zinotes.ui.viewmodel.AppViewModelProvider
import com.example.zinotes.ui.viewmodel.DetailViewModel

@Composable
fun DetailScreen(
    hanziId: Long?,
    onNavigateToEdit: (hanziId: Long) -> Unit,
    onNavigateBack: () -> Unit,
    modifier: Modifier = Modifier,
    viewModel: DetailViewModel = viewModel(factory = AppViewModelProvider.Factory)
) {
    LaunchedEffect(key1 = hanziId) {
        if (hanziId != null) {
            viewModel.loadHanzi(hanziId)
        }
    }

    val uiState by viewModel.uiState.collectAsState()

    var showDeleteDialog by remember { mutableStateOf(false) }

    if(showDeleteDialog) {
        ConfirmDialog(
            text = stringResource(R.string.confirm_delete_text),
            onDismiss = { showDeleteDialog = false },
            onConfirm = {
                showDeleteDialog = false
                viewModel.deleteHanzi(hanziId, onNavigateBack)
            }
        )
    }

    Box(
        modifier = modifier.fillMaxSize(),
        contentAlignment = Alignment.Center
    ) {
        Image(
            painter = painterResource(R.drawable.background),
            contentDescription = null,
            contentScale = ContentScale.FillBounds,
            modifier = Modifier.fillMaxSize()
        )
        Crossfade(targetState = uiState, label = "Content Crossfade") { state ->
            when (state) {
                is DetailUiState.Loading -> {
                    Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                        CircularProgressIndicator()
                    }
                }
                is DetailUiState.Error -> {
                    Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                        Text(text = state.message, color = MaterialTheme.colorScheme.error)
                    }
                }
                is DetailUiState.Success -> {
                    HanziDetailView(
                        hanzi = state.hanzi,
                        modifier = Modifier.fillMaxSize(),
                        onEditClick = onNavigateToEdit,
                        onDeleteClick = {
                            showDeleteDialog = true
                        }
                    )
                }
            }
        }
    }
}

@Composable
fun HanziDetailView(
    hanzi: Hanzi,
    onEditClick: (hanziId: Long) -> Unit,
    onDeleteClick: () -> Unit,
    modifier: Modifier = Modifier){
    Box(
        contentAlignment = Alignment.Center,
        modifier = modifier
    ){
        Image(
            painter = painterResource(R.drawable.vertical_scroll),
            contentDescription = null,
            modifier = Modifier.fillMaxSize()
        )
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.SpaceEvenly,
            modifier = Modifier
                .fillMaxHeight()
                .padding(top = 40.dp, bottom = 40.dp)
                .width(250.dp)
        ) {
            OutlinedTextField(
                value = hanzi.pinyin,
                onValueChange = {},
                label = { Text(text = "Pinyin",
                    textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth()
                ) },
                readOnly = true,
                textStyle = MaterialTheme.typography.bodyLarge.copy(
                    textAlign = TextAlign.Center
                )
            )
            OutlinedTextField(
                value = hanzi.tones.map{x -> x.toString()}.reduce { acc, s -> "$acc, $s" },
                onValueChange = {},
                label = { Text(text = "Tones",
                    textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth()
                ) },
                readOnly = true,
                textStyle = MaterialTheme.typography.bodyLarge.copy(
                    textAlign = TextAlign.Center
                )
            )
            OutlinedTextField(
                value = hanzi.radicalNumber?.toString()?:"",
                onValueChange = {},
                label = { Text(text = "Radical Number",
                    textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth()
                ) },
                readOnly = true,
                textStyle = MaterialTheme.typography.bodyLarge.copy(
                    textAlign = TextAlign.Center
                )
            )
            OutlinedTextField(
                value = hanzi.strokeCount?.toString()?:"",
                onValueChange = {},
                label = { Text(text = "Stroke Count",
                    textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth()
                ) },
                readOnly = true,
                textStyle = MaterialTheme.typography.bodyLarge.copy(
                    textAlign = TextAlign.Center
                )
            )
            OutlinedTextField(
                value = hanzi.hskLevel?.toString()?:"",
                onValueChange = {},
                label = { Text(text = "HSK Level",
                    textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth()
                ) },
                readOnly = true,
                textStyle = MaterialTheme.typography.bodyLarge.copy(
                    textAlign = TextAlign.Center
                )
            )
            OutlinedTextField(
                value = hanzi.definitions?.reduce { acc, s -> "$acc, $s" }?:"",
                onValueChange = {},
                label = { Text(text = "Definitions",
                    textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth()
                ) },
                readOnly = true,
                textStyle = MaterialTheme.typography.bodyLarge.copy(
                    textAlign = TextAlign.Center
                )
            )
            Row(
                horizontalArrangement = Arrangement.SpaceBetween,
                modifier = Modifier.fillMaxWidth().padding(start = 16.dp, end = 16.dp)
            ) {
                IconButton(
                    onClick = onDeleteClick,
                    colors = IconButtonDefaults.iconButtonColors(
                        containerColor = MaterialTheme.colorScheme.errorContainer
                    )
                ) {
                    Icon(
                        imageVector = Icons.Filled.Delete,
                        contentDescription = null,
                        tint = MaterialTheme.colorScheme.onSurface
                    )
                }
                IconButton(
                    onClick = {
                        onEditClick(hanzi.id)
                    },
                    colors = IconButtonDefaults.iconButtonColors(
                        containerColor = MaterialTheme.colorScheme.primaryContainer
                    )
                ) {
                    Icon(
                        imageVector = Icons.Filled.Edit,
                        contentDescription = null,
                        tint = MaterialTheme.colorScheme.onSurface
                    )
                }
            }
        }
    }
}

File: non-native/zinotes_flutter/lib/screen/detail_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import '../model/hanzi.dart';
import '../notifier/list_notifier.dart';
import 'confirm_dialog.dart';
import 'edit_screen.dart';

class DetailScreen extends StatelessWidget {
  final String hanziId;

  const DetailScreen({super.key, required this.hanziId});

  @override
  Widget build(BuildContext context) {
      return Scaffold(
          body: Container(
            decoration: BoxDecoration(image: DecorationImage(image: AssetImage('assets/images/background.jpeg'))),
            child: HanziDetailView(hanziId)
          )
      );
  }
}

class HanziDetailView extends StatelessWidget{
  final String hanziId;

  const HanziDetailView(this.hanziId, {super.key});

  @override
  Widget build(BuildContext context) {
    final notifier = context.watch<ListNotifier>();
    Hanzi? hanzi = notifier.getHanziById(hanziId);

    if (hanzi == null) {
      return Text('Hanzi not found');
    }

    final ValueKey<Hanzi> key = ValueKey<Hanzi>(hanzi);

    return Container(
      key: key,
      decoration: BoxDecoration(image: DecorationImage(
        image: AssetImage('assets/images/vertical scroll.png'),
        fit: BoxFit.fitHeight,
      )),
      child: Padding(
        padding: const EdgeInsets.fromLTRB(80,20,80,20),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            TextFormField(
              decoration: InputDecoration(
                floatingLabelAlignment: FloatingLabelAlignment.center,
                border: OutlineInputBorder(),
                labelText: 'Pinyin',
                alignLabelWithHint: true,
              ),
              initialValue: hanzi.pinyin,
              readOnly: true,
              textAlign: TextAlign.center,
              style: Theme.of(context).textTheme.bodyLarge,
            ),
            TextFormField(
              decoration: InputDecoration(
                floatingLabelAlignment: FloatingLabelAlignment.center,
                border: OutlineInputBorder(),
                labelText: 'Tones',
                alignLabelWithHint: true,
              ),
              initialValue: hanzi.tones.join(','),
              readOnly: true,
              textAlign: TextAlign.center,
              style: Theme.of(context).textTheme.bodyLarge,
            ),
            TextFormField(
              decoration: InputDecoration(
                floatingLabelAlignment: FloatingLabelAlignment.center,
                border: OutlineInputBorder(),
                labelText: 'Radical Number',
                alignLabelWithHint: true,
              ),
              initialValue: hanzi.radicalNumber?.toString(),
              readOnly: true,
              textAlign: TextAlign.center,
              style: Theme.of(context).textTheme.bodyLarge,
            ),
            TextFormField(
              decoration: InputDecoration(
                floatingLabelAlignment: FloatingLabelAlignment.center,
                border: OutlineInputBorder(),
                labelText: 'Stroke Count',
                alignLabelWithHint: true,
              ),
              initialValue: hanzi.strokeCount?.toString(),
              readOnly: true,
              textAlign: TextAlign.center,
              style: Theme.of(context).textTheme.bodyLarge,
            ),
            TextFormField(
              decoration: InputDecoration(
                floatingLabelAlignment: FloatingLabelAlignment.center,
                border: OutlineInputBorder(),
                labelText: 'HSK Level',
                alignLabelWithHint: true,
              ),
              initialValue: hanzi.hskLevel?.toString(),
              readOnly: true,
              textAlign: TextAlign.center,
              style: Theme.of(context).textTheme.bodyLarge,
            ),
            TextFormField(
              decoration: InputDecoration(
                floatingLabelAlignment: FloatingLabelAlignment.center,
                border: OutlineInputBorder(),
                labelText: 'Definitions',
                alignLabelWithHint: true,
              ),
              initialValue: hanzi.definitions?.join(','),
              readOnly: true,
              textAlign: TextAlign.center,
              style: Theme.of(context).textTheme.bodyLarge,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                IconButton(
                  icon: Icon(Icons.delete),
                  onPressed: () {
                    showDialog(
                      context: context,
                      barrierDismissible: true,
                      builder: (BuildContext context) {
                        return ConfirmDialog(hanziId: hanzi.id);
                      },
                    );
                  },
                  color: Colors.red,
                ),
                IconButton(
                    icon: Icon(Icons.edit),
                    onPressed: () {
                      Navigator.push(
                        context,
                        MaterialPageRoute(builder: (context) => EditScreen(hanziId: hanzi.id)),
                      );
                    }
                )
              ]
            )
          ]
        ),
      )
    );
  }
}